chg: dev: Modified iperf libraries

herjosep opened the pull request:
Added multiple instances support so multiple clients or servers can run at the same time

Commit:
chg: dev: Modified iperf libraries

Added multiple instances support so multiple clients or servers can run at the same time

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -31,24 +31,32 @@ class IperfServerState(object):
-        self.server_pid = server_pid
+
+    def __init__(self, server_pids={}):
+        self.server_pids = server_pids
```
Warning! Never use mutable structures as default keyword arguments. Ever!

Check this:

http://reinout.vanrees.org/weblog/2012/04/18/default-parameters.html

Use instead:

``` python
def __init__(self, server_pids=None):
    self.server_pids = server_pids or {}
```

Or the more verbose:

``` python
def __init__(self, server_pids=None):
    if server_pids is None:
        server_pids = {}
    self.server_pids = server_pids
```

carlos-jenkins replied:
Also, given the fact that you aren't passing an initialization function to the stateprovider decorator, you can simply remove the keyword argument altogether and just leave the attribute.

``` python
def __init__(self):
    self.server_pids = {}
```

herjosep replied:
Ok, I'll fix it and not use any argument.

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -31,24 +31,32 @@ class IperfServerState(object):
-    def __init__(self, client_pid=None):
-        self.client_pid = client_pid
+
+    def __init__(self, client_pids={}):
```
Same here.

herjosep replied:
Ok

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -31,24 +31,32 @@ class IperfServerState(object):
+        clients instances as defined by {instance_id: client_pid} where
+     :param int instance_id: Instance of iperf server and
+     :param int client_pid: PID number of iperf server instance.
     """
```
Please correctly format the docstring and pass the `tox -e doc` to ensure it correctness.

herjosep replied:
Ok

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -57,6 +65,7 @@ def server_start(enode, state, port, interval=1, udp=False):
     :param int port: iperf port to be open
     :param int interval: interval for iperf server to check
     :param bool udp: If it is UDP or TCP. Default is False for TCP.
+    :param int instance_id: Number of iperf server instance    .
```
Why that many whitespaces?

herjosep replied:
That is most likely a weird typo on my part. I'll fix it.

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -67,34 +76,45 @@ def server_start(enode, state, port, interval=1, udp=False):
-    state.server_pid = parse_pid(enode(' '.join(cmd), shell='bash'))
+    state.server_pids[instance_id] = parse_pid(enode(' '.join(cmd),
+                                                     shell='bash'))
 
```
Semantics are wrong. You're defining `instance_id` as a keyword argument. This implies that it is optional. Nevertheless, you enforce it as a key for the `server_pids` dictionary. You're assuming that the default value, `None`, will be used as key? Can you just not assume that the default should be, don't know, `1`?

If you do so, semantics gets right, and you can just, for example:

``` python
cmd.append('2>&1 > /tmp/iperf_server-{} &'.format(instance_id))
```

herjosep replied:
None is a perfectly valid key for a dictionary in Python and my intent was that None is the default key to preserve original functionality as much as possible.
It's something I can change of course to simplify lines of code.

carlos-jenkins replied:
Great, yes, to simplify lines of code, but also because it is also a good practice to not mix different datatypes in a dictionary for both keys and values. Consider:

``` python
mydict = {
    1: 'a',
    None: 'b'
}

sum(mydict.keys())  # <- Crash
```

Same applies with values. It is good practice to have a consistent key datatype and data datatype in the dictionary.

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -67,34 +76,45 @@ def server_start(enode, state, port, interval=1, udp=False):
 
-    state.server_pid = parse_pid(enode(' '.join(cmd), shell='bash'))
+    state.server_pids[instance_id] = parse_pid(enode(' '.join(cmd),
+                                                     shell='bash'))
```
It is a good practice to relieve horizontal space so subsequent commands can use more horizontal space. For example:

``` python
state.server_pids[instance_id] = parse_pid(
  enode(' '.join(cmd), shell='bash')
)
```

herjosep replied:
I'll fix it.

Commit:
fix: dev: Modified iperf libraries

Modified iperf libraries according to comments done during pull request to main branch

Commit:
fix: dev: Modified iperf libraries

Text review for all functions in library.py for changes requested in pull-request.

Commit:
chg: dev: Modified iperf libraries

- Added extra help information for functions
- Added bandwidth setting to client_start

Commit:
fix: dev: Fixed documentation indents

Fixed documentation indents for documentation to match that found in other libraries

Commit:
chg: dev: Modified shell support on iperf libraries

Changed the iperf libraries to accept different shells.
This will make iperf be able to execute on devices that are not using bash as shell.

herjosep commented:
All the comments are corrected.
Plus, added 2 additional things that were required recently for the methods to support:
- Bandwidth definition
- Different shell support

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -102,11 +120,18 @@ def client_start(
+    :param int port: iperf port to be open.
+    :param int interval: interval for iperf server to check.
+    :param int time: the time iperf client will be running.
     :param bool udp: If it is UDP or TCP. Default is False for TCP.
```
That's not true. Currently default is `None`. 

Most of the time we set the default to be the same of the expected datatype of the attribute. And we put the default as the keyword argument value.

herjosep replied:
Will change to udp=False to match the other functions. The default comment I think still adds value here because the oppositve of UDP (TCP) may not be so obvious in meaning.

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -102,11 +120,18 @@ def client_start(
     :param bool udp: If it is UDP or TCP. Default is False for TCP.
+    :param str bandwidth: Bandwidth for iperf to use in bits/sec.
+     When used automatically switches to UDP regardless of udp setting.
+     Default is 1Mb/sec.
```
Same here. I would remote the "Default is 1Mb/sec." and just put in the signature: `bandwidth='1Mb/sec'`.

herjosep replied:
What I'm going to do here is remove the default note, which is the default for iperf. If I set the default as part of the argument iperf will start malfunctioning since the bandwidth option switches automatically to UDP regardless of what the user has set.

herjosep replied:
I'm going to note here instead that iperf by default transmits at 1Mb/sec on either TCP and UDP.

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -30,33 +30,38 @@
+def server_start(
+                 enode, state, port, interval=1,
+                 udp=False, instance_id=1, shell=None
+                 ):
```
The `client_start` function has a different indentation style. Consider the following:

``` python
                                                    Assume this is line 80
                                                           |
                                                           v
def this_is_a_very_long_long_function_name(
                                           arg1, 
                                           arg2,
                                           long_attribute_name
                                           ):
    foo = 100
```

Versus:

``` python
def this_is_a_very_long_long_function_name(
        arg1, arg2,
        long_attribute_name):
    foo = 100
```

The former will allow you to reset the horizontal space for subsequent lines.

herjosep replied:
I will correct it.

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -102,11 +120,18 @@ def client_start(
+     When used automatically switches to UDP regardless of udp setting.
+     Default is 1Mb/sec.
+    :param int instance_id: Number of iperf client instance.
+     Default is 1.
```
Same here. If we have the default in the constructor it is advised to not put it here.

herjosep replied:
I'm going to remove it.

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -120,33 +145,47 @@ def client_start(
     :param enode: Engine node to communicate with.
     :type enode: topology.platforms.base.BaseNode
+    :param int instance_id: Number of iperf client instance.
+     Default is 1.
```
Same here.

herjosep replied:
Same, going to remove it.

carlos-jenkins commented:
That would be my final observations. The implementation is fine, simple and straightforward. Great job.

Commit:
fix: dev: Modified iperf libraries

- Fixed documentation for several functions
- Modified default setting for udp setting on client_start function on library.py
- Fixed identation on server_start and client_Start functions on library.py

herjosep commented:
Fixed the observations you mentioned. Thanks.

Commit:
chg: dev: Modified decorator of iperf libraries

Modified the number of state classes to just 1. Testing and documentation confirm only 1 decorator is allowed per library on the current topology implementation.
Currently, if iperf is run as client and then run a server function for example on the same device; the function fails because the decorator is set for the client state, not the server one.

herjosep commented:
One last change I had to implement. This will allow the library to support a device that has to run both iperf server and client processes.

carlos-jenkins commented on lib/topology_lib_iperf/library.py:
```diff
@@ -107,12 +102,19 @@ def server_stop(enode, state, instance_id=1, shell=None):
+        bandwidth=None,
+        instance_id=1,
+        shell=None
+):
```
I preferred way better the previous version here. Yes, you reset the horizontal space, but now you're wasting vertical space :( Anyhow, that's just me being picky.

carlos-jenkins commented:
> Modified the number of state classes to just 1. Testing and documentation confirm only 1 decorator is allowed per library on the current topology implementation.
> Currently, if iperf is run as client and then run a server function for example on the same device; the function fails because the decorator is set for the client state, not the server one.

While I love the changes, and make a lot of sense to have only one state class for this library, I think that the current implementation **do** allow multiple state classes to be used. You say:

> Testing and documentation confirm only 1 decorator is allowed per library on the current topology implementation.

Which documentation? Nor the [Plugin development guide](http://topology.readthedocs.org/en/latest/plugins.html#saving-state), nor the [stateprovider reference documentation](http://topology.readthedocs.org/en/latest/topology/topology.libraries.utils.html#topology.libraries.utils.stateprovider), not even the [implementation of the decorator](https://github.com/HPENetworking/topology/blob/master/lib/topology/libraries/utils.py#L64) shows that only one class can be used.

If you did found a bug I'll love to see the traceback to understand it. AFAIK, that limitation doesn't exists.

carlos-jenkins commented:
@herjosep I'm merging this. Its great, I was just expecting answer for the above questions.
